File: src/AttributeMap.ts
// Copyright (c) 2022, Slab, Inc.
// Copyright (c) 2026, RealtimeBoard, Inc. dba Miro
// SPDX-License-Identifier: BSD-3-Clause

// eslint-disable
import { isEqual } from 'es-toolkit';

const MAX_RECURSION_DEPTH = 100;

export interface AttributeMap {
  [key: string]: unknown;
}

function isNestedMap(value: unknown): value is AttributeMap {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function safeKeys(map: AttributeMap): string[] {
  return Object.keys(map).filter((key) => key !== '__proto__');
}

/**
 * Thrown instead of silently truncating a clone or merge, which would either leak an
 * unfiltered (and potentially __proto__-carrying) subtree past the depth budget, or make
 * replicas converge on different documents.
 */
export class NestingDepthExceededError extends Error {
  constructor() {
    super(`AttributeMap nesting exceeds the maximum depth`);
    this.name = 'NestingDepthExceededError';
  }
}

/**
 * Clones the attribute map while also;
 *  - removing unsafe __proto__ keys
 *  - ensuring it doesn't exceed depth budget
 *  - stripping nulls if requested
 */
function sanitizeClone(
  attr: AttributeMap,
  depth: number,
  keepNull: boolean,
): AttributeMap | undefined {
  const attributes: AttributeMap = {};
  for (const key of safeKeys(attr)) {
    const value = attr[key];
    if (isNestedMap(value)) {
      if (depth <= 1) {
        throw new NestingDepthExceededError();
      }
      const res = sanitizeClone(value, depth - 1, keepNull);
      if (res === undefined) {
        continue;
      } else {
        attributes[key] = res;
      }
    } else if (keepNull || value != null) {
      attributes[key] = value;
    }
  }

  return keepNull || Object.keys(attributes).length > 0 ? attributes : undefined;
}

export namespace AttributeMap {
  export function compose(
    a: AttributeMap = {},
    b: AttributeMap = {},
    keepNull = false,
    depth = MAX_RECURSION_DEPTH,
  ): AttributeMap | undefined {
    if(depth<=1){
      throw new NestingDepthExceededError()
    }
    if (typeof a !== 'object') {
      a = {};
    }
    if (typeof b !== 'object') {
      b = {};
    }
    const attributes = sanitizeClone(b, depth, keepNull) ?? {};

    safeKeys(a).forEach((key) => {
      if (isNestedMap(a[key]) && isNestedMap(b[key]) && depth > 1) {
        const nestedComposed = AttributeMap.compose(
          a[key] as AttributeMap,
          b[key] as AttributeMap,
          keepNull,
          depth - 1,
        );
        if (nestedComposed === undefined) {
          delete attributes[key];
        } else {
          attributes[key] = nestedComposed;
        }
      } else if (a[key] !== undefined && b[key] === undefined) {
        attributes[key] = a[key];
      }
    });

    return Object.keys(attributes).length > 0 ? attributes : undefined;
  }

  export function diff(
    a: AttributeMap = {},
    b: AttributeMap = {},
    depth = MAX_RECURSION_DEPTH,
  ): AttributeMap | undefined {
    if (typeof a !== 'object') {
      a = {};
    }
    if (typeof b !== 'object') {
      b = {};
    }
    const attributes = safeKeys(structuredClone(a))
      .concat(safeKeys(structuredClone(b)))
      .reduce<AttributeMap>((attrs, key) => {
        if (!isEqual(a[key], b[key])) {
          if (isNestedMap(a[key]) && isNestedMap(b[key]) && depth > 1) {
            const nestedDiff = AttributeMap.diff(
              a[key] as AttributeMap,
              b[key] as AttributeMap,
              depth - 1,
            );
            if (nestedDiff !== undefined) {
              attrs[key] = nestedDiff;
            }
          } else {
            attrs[key] = b[key] === undefined ? null : b[key];
          }
        }
        return attrs;
      }, {});
    return Object.keys(attributes).length > 0 ? attributes : undefined;
  }

  export function invert(
    attr: AttributeMap = {},
    base: AttributeMap = {},
    depth = MAX_RECURSION_DEPTH,
  ): AttributeMap {
    attr = attr || {};
    const baseInverted = safeKeys(base).reduce<AttributeMap>((memo, key) => {
      if (!isEqual(base[key], attr[key]) && attr[key] !== undefined) {
        if (isNestedMap(base[key]) && isNestedMap(attr[key]) && depth > 1) {
          const nested = AttributeMap.invert(
            attr[key] as AttributeMap,
            base[key] as AttributeMap,
            depth - 1,
          );
          if (Object.keys(nested).length > 0) {
            memo[key] = nested;
          }
        } else {
          memo[key] = base[key];
        }
      }
      return memo;
    }, {});
    return safeKeys(attr).reduce<AttributeMap>((memo, key) => {
      if (!isEqual(attr[key], base[key]) && base[key] === undefined) {
        memo[key] = null;
      }
      return memo;
    }, baseInverted);
  }

  export function transform(
    a: AttributeMap | undefined,
    b: AttributeMap | undefined,
    priority = false,
    depth = MAX_RECURSION_DEPTH,
  ): AttributeMap | undefined {
    if (typeof a !== 'object') {
      return b;
    }
    if (typeof b !== 'object') {
      return undefined;
    }
    if (!priority) {
      return b; // b is unchanged when a doesn't have priority
    }
    const attributes = safeKeys(b).reduce<AttributeMap>((attrs, key) => {
      if (isNestedMap(a[key]) && isNestedMap(b[key]) && depth > 1) {
        const attr = AttributeMap.transform(
          a[key] as AttributeMap,
          b[key] as AttributeMap,
          priority,
          depth - 1,
        );
        if (attr !== undefined) {
          attrs[key] = attr;
        }
      } else if (a[key] === undefined) {
        attrs[key] = b[key]; // null is a valid value
      }
      return attrs;
    }, {});
    return Object.keys(attributes).length > 0 ? attributes : undefined;
  }
}

File: languages/python/src/rich_text_delta/attribute_map.py
# Copyright (c) 2022, Slab, Inc.
# Copyright (c) 2026, RealtimeBoard, Inc. dba Miro
# SPDX-License-Identifier: BSD-3-Clause

"""Attribute maps and the four operations that combine them.

Attribute values may themselves be maps; ``compose``, ``diff``, ``invert`` and
``transform`` recurse into them up to ``MAX_RECURSION_DEPTH`` levels instead of treating
them as scalars. A value of ``None`` is a removal ("null" on the wire); a key that is
absent is simply unmentioned.
"""

from __future__ import annotations

from copy import deepcopy
from typing import Any, Dict, List, Optional

from ._js import UNDEFINED, deep_equal, prop

AttributeMap = Dict[str, Any]

MAX_RECURSION_DEPTH = 100


def _is_nested_map(value: Any) -> bool:
    return isinstance(value, dict)


def _safe_keys(map_: AttributeMap) -> List[str]:
    return [key for key in map_ if key != '__proto__']


def compose(
    a: Optional[AttributeMap] = None,
    b: Optional[AttributeMap] = None,
    keep_null: bool = False,
    depth: int = MAX_RECURSION_DEPTH,
) -> Optional[AttributeMap]:
    """Apply ``b`` on top of ``a``, returning ``None`` when nothing is left.

    Unless ``keep_null``, keys whose value is ``None`` are dropped rather than recorded as
    removals.
    """
    if not isinstance(a, dict):
        a = {}
    if not isinstance(b, dict):
        b = {}
    attributes = deepcopy(b)
    if not keep_null:
        attributes = {
            key: attributes[key] for key in _safe_keys(attributes) if attributes[key] is not None
        }
    for key in _safe_keys(a):
        if _is_nested_map(a.get(key)) and _is_nested_map(b.get(key)) and depth > 1:
            nested_composed = compose(a[key], b[key], keep_null, depth - 1)
            if nested_composed is None:
                attributes.pop(key, None)
            else:
                attributes[key] = nested_composed
        elif prop(a, key) is not UNDEFINED and prop(b, key) is UNDEFINED:
            attributes[key] = a[key]
    return attributes if len(attributes) > 0 else None


def diff(
    a: Optional[AttributeMap] = None,
    b: Optional[AttributeMap] = None,
    depth: int = MAX_RECURSION_DEPTH,
) -> Optional[AttributeMap]:
    """The change that turns ``a`` into ``b``, or ``None`` when they already match."""
    if not isinstance(a, dict):
        a = {}
    if not isinstance(b, dict):
        b = {}
    attrs: AttributeMap = {}
    for key in _safe_keys(a) + _safe_keys(b):
        if not deep_equal(prop(a, key), prop(b, key)):
            if _is_nested_map(a.get(key)) and _is_nested_map(b.get(key)) and depth > 1:
                nested_diff = diff(a[key], b[key], depth - 1)
                if nested_diff is not None:
                    attrs[key] = nested_diff
            else:
                attrs[key] = None if prop(b, key) is UNDEFINED else b[key]
    return attrs if len(attrs) > 0 else None


def invert(
    attr: Optional[AttributeMap] = None,
    base: Optional[AttributeMap] = None,
    depth: int = MAX_RECURSION_DEPTH,
) -> AttributeMap:
    """The change that undoes ``attr`` when it was applied to ``base``."""
    if attr is None:
        attr = {}
    if base is None:
        base = {}
    base_inverted: AttributeMap = {}
    for key in _safe_keys(base):
        if not deep_equal(prop(base, key), prop(attr, key)) and prop(attr, key) is not UNDEFINED:
            if _is_nested_map(base.get(key)) and _is_nested_map(attr.get(key)) and depth > 1:
                nested = invert(attr[key], base[key], depth - 1)
                if len(nested) > 0:
                    base_inverted[key] = nested
            else:
                base_inverted[key] = base[key]
    memo = base_inverted
    for key in _safe_keys(attr):
        if not deep_equal(prop(attr, key), prop(base, key)) and prop(base, key) is UNDEFINED:
            memo[key] = None
    return memo


def transform(
    a: Optional[AttributeMap],
    b: Optional[AttributeMap],
    priority: bool = False,
    depth: int = MAX_RECURSION_DEPTH,
) -> Optional[AttributeMap]:
    """Rewrite ``b`` so it can be applied after ``a``.

    Without ``priority``, ``b`` wins outright and is returned unchanged.
    """
    if not isinstance(a, dict):
        return b
    if not isinstance(b, dict):
        return None
    if not priority:
        return b  # b is unchanged when a doesn't have priority
    attributes: AttributeMap = {}
    for key in _safe_keys(b):
        if _is_nested_map(a.get(key)) and _is_nested_map(b.get(key)) and depth > 1:
            attr = transform(a[key], b[key], priority, depth - 1)
            if attr is not None:
                attributes[key] = attr
        elif prop(a, key) is UNDEFINED:
            attributes[key] = b[key]  # None is a valid value
    return attributes if len(attributes) > 0 else None
